File: leet_code/trees/build_tree_pre_inorder.rb
class TreeNode
  attr_accessor :val, :left, :right
  def initialize(val)
    @val = val
    @left = nil
    @right = nil
  end
end

def build_tree(preorder, inorder)
    return nil if preorder.empty?

    idx = inorder.index(preorder[0])
    new_node = TreeNode.new(preorder[0])

    new_node.left = build_tree(preorder[1..idx], inorder[0...idx])
    new_node.right = build_tree(preorder[idx+1..-1], inorder[idx+1..-1])
    new_node
end

File: leet_code/trees/build_tree_pre_inorder.js
class TreeNode{
  constructor(val){
    this.val = val;
    this.left = null;
    this.right = null;
  }
}

const buildTree = (preorder, inorder) => {
    if (preorder.length === 0) return null;

    const pLength = preorder.length,
          iLength = inorder.length,
          idx = inorder.indexOf(preorder[0]),
          newNode = new TreeNode(preorder[0]);

    newNode.left = buildTree(preorder.slice(1,idx+1), inorder.slice(0,idx));
    newNode.right = buildTree(preorder.slice(idx+1, pLength), inorder.slice(idx+1, iLength));
    return newNode;

};
